File: chapter_4_trees_graphs/js/binarySearchTree.js
class BinarySearchTree {
  /* Binary Search Tree constructor */
  constructor() {
    this.root = null
    this.size = 0
  }

  /* Inserts a node into a Binary Search Tree */
  insert(value) {
    const newNode = new TreeNode(value)
    if (!this.root) this.root = newNode
    else this.addNode(this.root, newNode)
    this.size++
    return newNode
  }

  /* Inserts a node onto another node */
  addNode(node, newNode) {
    if (newNode.value <= node.value) {
      /* insert in left side */
      if (!node.left) {
        /* if left node does not exist, insert */
        newNode.parent = node
        node.left = newNode
      } else {
        /* recurse */
        this.addNode(node.left, newNode)
      }
    } else {
      /* insert in right side */
      if (!node.right) {
        /* if right node does not exist, insert */
        newNode.parent = node
        node.right = newNode
      } else {
        /* recurse */
        this.addNode(node.right, newNode)
      }
    }
  }
}

module.exports = {
  BinarySearchTree
}

File: chapter_4_trees_graphs/python/binary_search_tree.py
TreeNode = __import__('binary_search_tree').BinarySearchTree


class BinarySearchTree:
    def __init__(self):
        """
        Binary Search Tree constructor
        """
        self.root = None
        self.size = 0

    def insert(self, value):
        """
        Inserts a node into a Binary Search Tree
        """
        node = TreeNode(value)
        if not self.root:
            self.root = node
        else:
            self.add_node(self.root, node)
        self.size += 1
        return node

    def add_node(self, node, new_node):
        """
        Inserts a node onto another node
        """
        if new_node.value <= node.value:
            # insert in left side
            if not node.left:
                # if left node does not exist, insert
                node.left = new_node
                new_node.parent = node
            else:
                # recurse
                self.add_node(node.left, new_node)
        else:
            # insert in right side
            if not node.right:
                # if right node does not exist, insert
                node.right = new_node
                new_node.parent = node
            else:
                # recurse
                self.add_node(node.right, new_node)
